### PR TITLE
[Review]zendesk-exporters-services (ticket-metrics table and  users table and ticket-comments table)

### PR DESCRIPTION
**Pull request by lanbo-phil**
https://app.asana.com/0/0/1134983042831949/f

### CODE REVIEW

File: zendesk/zd_ticket_metrics_service.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import (
+	"fmt"
+	"time"
+)
+
+// Ticket represents a Zendesk Ticket.
+//
+// Zendesk Core API docs: https://developer.zendesk.com/rest_api/docs/core/ticket_metrics
+
+type TicketMetric struct {
```
Comment (narup) [inline zendesk/zd_ticket_metrics_service.go]: Good work on including the link but first line of comment is misleading. Ticket or TicketMetrics?

File: zendesk/zd_ticket_metrics_service.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	StatusUpdatedAt      *time.Time `json:"status_updated_at,omitempty"`
+	InitiallyAssigneeAt  *time.Time `json:"initially_assigned_at,omitempty"`
+	AssignedAt           *time.Time `json:"assigned_at,omitempty"`
+	SolvedAt             *time.Time `json:"solved_at,omitempty"`
+	LatestCommentAddedAt *time.Time `json:"latest_comment_added_at,omitempty"`
+	FirstResolutionTime  Object     `json:"first_resolution_time_in_minutes,omitempty"`
+	ReplyTime            Object     `json:"reply_time_in_minutes,omitempty"`
+	FullResolutionTime   Object     `json:"full_resolution_time_in_minutes,omitempty"`
+
+	AgentWaitTime     Object `json:"agent_wait_time_in_minutes,omitempty"`
```
Comment (narup) [inline zendesk/zd_ticket_metrics_service.go]: remove the extra line gap

File: zendesk/zendesk_client_service.go [hunk 3]
```diff
@@ -101,6 +105,19 @@ func NewEnvClient(middleware ...MiddlewareFunction) (Client, error) {
 
 // NewClient creates a new Client.
 //
+
+// // NewZendeskEnvClient create zendesk client for request
+func NewEnvClient() (zClient *client, error) {
+	if zClient == nil {
+		c, err := zendesk.NewClient(util.Config("zendesk.domain"), util.Config("zendesk.username"), util.Config("zendesk.password"))
```
Comment (narup) [inline zendesk/zendesk_client_service.go]: hmm this won't work. util comes from plib zendesk is indepandant of plib. function should take all these as an argument

File: zendesk/zd_ticket_metrics_service.go (hunk 1 above)
Comment (lanbo-phil) [reply zendesk/zd_ticket_metrics_service.go]: Updated.

File: zendesk/zd_ticket_metrics_service.go (hunk 2 above)
Comment (lanbo-phil) [reply zendesk/zd_ticket_metrics_service.go]: Updated.

File: zendesk/zendesk_client_service.go (hunk 3 above)
Comment (lanbo-phil) [reply zendesk/zendesk_client_service.go]: This will be handled before the merge.

File: zendesk/zd_ticket_metrics_service.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+	defer res.Body.Close()
+
+	err = unmarshall(res, dataPerPage)
+
+	prevPage := ""
+	for dataPerPage.NextPage != prevPage {
+		result = append(result, dataPerPage.TicketMetrics...)
+		prevPage = dataPerPage.NextPage
+		res, _ := c.request("GET", dataPerPage.NextPage[38:], headers, bytes.NewReader(payload))
```
Comment (abhyuday) [inline zendesk/zd_ticket_metrics_service.go]: why is the index starting from 38? Some comment would help

File: zendesk/zd_ticket_metrics_service.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+type Object struct {
+	Calendar int64 `json:"calendar"`
+	Business int64 `json:"business"`
+}
+
+// func (c *client) GetAllTicketMetrics() ([]TicketMetric, error) {
+// 	out := new(APIPayload)
+// 	err := c.get("/api/v2/ticket_metrics.json", out)
+// 	return out.TicketMetrics, err
+// }
```
Comment (narup) [inline zendesk/zd_ticket_metrics_service.go]: Please remove dead code

Comment (lanbo-phil) [reply zendesk/zd_ticket_metrics_service.go]: Just cleaned.

File: zendesk/zd_ticket_metrics_service.go (hunk 4 above)
Comment (lanbo-phil) [reply zendesk/zd_ticket_metrics_service.go]: Put comments on after that  in the new commit.

File: zendesk/zd_ticket_comments_service.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	PlainBody   string       `json:"plain_body,omitempty"`
+	Public      bool         `json:"public,omitempty"`
+	AuthorID    int64        `json:"author_id,omitempty"`
+	Attachments []Attachment `json:"attachments,omitempty"`
+	Via         *Via         `json:"via,omitempty"`
+	MetaData    interface{}  `json:"metadata,omitempty"`
+	CreatedAt   *time.Time   `json:"created_at,omitempty"`
+	Uploads     []string     `json:"uploads,omitempty"`
+}
+
```
Comment (XiaokaiPhil) [inline zendesk/zd_ticket_comments_service.go]: we have comments and attachments defined in the **zd_ticket_service.go**, why you create a new one?

File: zendesk/zendesk_client_service.go [hunk 7]
```diff
@@ -241,7 +241,11 @@ func (c *client) getAll(endpoint string, in interface{})([]Ticket, error) {
 	for dataPerPage.NextPage != prevPage {
 		result = append(result, dataPerPage.Tickets...)
 		prevPage = dataPerPage.NextPage
+		if prevPage == "" {
+			break
+		}
```
Comment (XiaokaiPhil) [inline zendesk/zendesk_client_service.go]: the PR to fix this function is merged. You can pull the laster version to get the updated function.

File: zendesk/zd_ticket_comments_service.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+/*
+//GET /api/v2/tickets/{ticket_id}/comments.json
+
+curl https://{subdomain}.zendesk.com/api/v2/tickets/{ticket_id}/comments.json \
+-H "Content-Type: application/json" -v -u {email_address}:{password}
+*/
+
+func (c *client) ShowTicketComments(id int64) ([]TicketComment, error) {
+	out := new(APIPayload)
```
Comment (XiaokaiPhil) [inline zendesk/zd_ticket_comments_service.go]: what is the difference between this function and ListTicketComments?

File: zendesk/zd_ticket_comments_service.go (hunk 6 above)
Comment (lanbo-phil) [reply zendesk/zd_ticket_comments_service.go]: If you read closely of https://developer.zendesk.com/rest_api/docs/support/tickets
you will see "comments" and "attachments" are not supposed in ticket struct. 
That is the first reason why you see the decoupled ticket_comments_service.go

Another more important reason is to align with the sql table structure, which naturally requires decoupling "comments table" and "ticket table" for performance consideration.

File: zendesk/zendesk_client_service.go (hunk 7 above)
Comment (lanbo-phil) [reply zendesk/zendesk_client_service.go]: This commit is before your merge.

File: zendesk/zd_ticket_comments_service.go (hunk 8 above)
Comment (lanbo-phil) [reply zendesk/zd_ticket_comments_service.go]: These two functions are doing the same but in different service.

File: zendesk/zd_ticket_comments_service.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
+package zendesk
+
+import (
+	"fmt"
+	"time"
+)
+
+// Ticket represents a Zendesk Ticket.
```
Comment (narup) [inline zendesk/zd_ticket_comments_service.go]: Fix the comment

File: zendesk/zd_ticket_metrics_service.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+/* The following implementation works for no pagination case
+
+func (c *client) GetAllTicketMetrics() ([]TicketMetric, error) {
+	out := new(APIPayload)
+	err := c.get("/api/v2/ticket_metrics.json", out)
+	return out.TicketMetrics, err
+}
+
+*/
```
Comment (narup) [inline zendesk/zd_ticket_metrics_service.go]: delete dead code

File: zendesk/zd_ticket_service.go [hunk 11]
```diff
@@ -14,39 +14,48 @@ import (
...
-	AdditionalTags []string `json:"additional_tags,omitempty"`
-	RemoveTags     []string `json:"remove_tags,omitempty"`
+	ID          int64  `json:"id,omitempty"`
+	URL         string `json:"url,omitempty"`
+	ExternalID  string `json:"external_id,omitempty"`
+	Type        string `json:"type,omitempty"`
+	Subject     string `json:"subject,omitempty"`
+	RawSubject  string `json:"raw_subject,omitempty"`
+	Description string `json:"description,omitempty"`
+	// Comment          *TicketComment `json:"comment,omitempty"`
```
Comment (narup) [inline zendesk/zd_ticket_service.go]: delete this!

File: zendesk/zd_ticket_service.go [hunk 12]
```diff
@@ -122,36 +140,24 @@ func (c *client) DeleteTicket(id int64) error {
...
+// 	ID          int64        `json:"id,omitempty"`
+// 	Type        string       `json:"type,omitempty"`
+// 	Body        string       `json:"body,omitempty"`
+// 	HTMLBody    string       `json:"html_body,omitempty"`
+// 	Public      bool         `json:"public"`
+// 	AuthorID    int64        `json:"author_id,omitempty"`
+// 	Attachments []Attachment `json:"attachments,omitempty"`
+// 	CreatedAt   *time.Time   `json:"created_at,omitempty"`
+// 	Uploads     []string     `json:"uploads,omitempty"`
+// }
```
Comment (narup) [inline zendesk/zd_ticket_service.go]: delete

File: zendesk/zd_ticket_comments_service.go (hunk 9 above)
Comment (lanbo-phil) [reply zendesk/zd_ticket_comments_service.go]: Fixed.

File: zendesk/zd_ticket_metrics_service.go (hunk 10 above)
Comment (lanbo-phil) [reply zendesk/zd_ticket_metrics_service.go]: Deleted.

File: zendesk/zd_ticket_service.go (hunk 11 above)
Comment (lanbo-phil) [reply zendesk/zd_ticket_service.go]: Deleted.

File: zendesk/zd_ticket_service.go (hunk 12 above)
Comment (lanbo-phil) [reply zendesk/zd_ticket_service.go]: Deleted.

File: zendesk/zd_ticket_comments_service.go [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+//
+// Zendesk Core API docs: https://developer.zendesk.com/rest_api/docs/core/ticket_comments
+
+type TicketComment struct {
+	ID          int64        `json:"id,omitempty"`
+	Type        string       `json:"type,omitempty"`
+	Body        string       `json:"body,omitempty"`
+	HTMLBody    string       `json:"html_body,omitempty"`
+	PlainBody   string       `json:"plain_body,omitempty"`
+	Public      bool         `json:"public,omitempty"`
```
Comment (XiaokaiPhil) [inline zendesk/zd_ticket_comments_service.go]: Any reason to add **omitempty** in this **Public**?

Comment (lanbo-phil) [reply zendesk/zd_ticket_comments_service.go]: This is to omit the field of this particular struct to appear in json when it is empty.

**Pull request merged**